pkg/agent/multicluster/pod_route_controller.go

```diff
@@ -0,0 +1,444 @@
+	pod, isPod := obj.(*corev1.Pod)
+	if !isPod {
+		klog.ErrorS(errors.New("received unexpected object"), "enqueuePod can't process event", "obj", obj)
+		return
+	}
+
+	podIP := pod.Status.PodIP
+	if podIP == "" {
+		return
+	}
+
+	podName := types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name}
+	klog.InfoS("Processing Pod", "Pod", podName.String(), "podIP", podIP)
+
+	podSet := c.getPodsByIP(podIP)
+	if podSet == nil && isDelete {
+		return
+	}
+
+	if podSet != nil && isDelete {
+		if podSet.Has(podName.String()) {
+			if podSet.Len() == 1 {
+				c.deletePodNodeIP(podIP)
+			}
+			c.deletePodCache(podIP, podName)
+
+			c.podQueue.Add(&podNodeIPs{
+				podIP: podIP,
+			})
+		}
+		return
+	}
+
+	isValidPod := c.validatePod(pod)
+	if !isValidPod {
+		return
+	}
+
+	podNodeIP := pod.Status.HostIP
+	c.addPodCache(podIP, podName)
+	c.addOrUpdatePodNodeIPs(podIP, podNodeIP)
+	c.podQueue.Add(&podNodeIPs{podIP: podIP, podNodeIP: podNodeIP})
+}
+
+func (c *MCPodRouteController) Run(stopCh <-chan struct{}) {
+	defer c.gwQueue.ShutDown()
+	defer c.podQueue.ShutDown()
+
+	klog.InfoS("Starting controller", "controller", podRouteControllerName)
+	defer klog.InfoS("Shutting down controller", "controller", podRouteControllerName)
+	if !cache.WaitForNamedCacheSync(podRouteControllerName, stopCh, c.gwInformer.HasSynced) {
+		return
+	}
+	// Run a single routine to handle Gateway events.
+	go wait.Until(c.gatewayWorker, time.Second, stopCh)
+
+	// Run a single routine to clean up stale Pod flows.
+	go wait.Until(c.cleanUpStalePodFlows, 5*time.Second, stopCh)
+	<-stopCh
+}
+
+func (c *MCPodRouteController) gatewayWorker() {
+	for c.processGatewayNextWorkItem() {
+	}
+}
+
+func (c *MCPodRouteController) processGatewayNextWorkItem() bool {
+	key, quit := c.gwQueue.Get()
+	if quit {
+		return false
+	}
+	defer c.gwQueue.Done(key)
+
+	if k, ok := key.(string); !ok {
+		c.gwQueue.Forget(k)
+		klog.InfoS("Expected string in work queue but got", "object", k)
+		return true
+	} else if err := c.syncGateway(); err == nil {
+		c.gwQueue.Forget(key)
+	} else {
+		c.gwQueue.AddRateLimited(key)
+		klog.ErrorS(err, "Error syncing Gateway, requeuing", "key", key)
+	}
+	return true
+}
+
+func (c *MCPodRouteController) syncGateway() error {
+	activeGW, err := getActiveGateway(c.gwLister, c.namespace)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get an active Gateway")
+		return err
+	}
+
+	c.podWorkersStartedMutex.Lock()
+	defer c.podWorkersStartedMutex.Unlock()
+	// Stop Pod flow controller and clean up all installed Multi-cluster Pod flows,
+	// if the Node was a Gateway before.
+	if activeGW == nil || c.nodeConfig.Name != activeGW.Name {
+		if c.podWorkersStarted {
+			klog.InfoS("Shutting down Multi-cluster PodFlowController")
+			close(c.podWorkerStopCh)
+			c.podInformer = nil
+			c.podLister = nil
+			c.podWorkersStarted = false
+		}
+		if len(c.podCache) > 0 {
+			err := c.ofClient.UninstallMulticlusterPodFlows("")
```

Comment:
As this may fail, and at retry the Node can become the Gateway again, we need to handle the cleanup in that case too. So, I suggest the following at the beginning of this func:

```
                isGateway := activeGW != nil && c.nodeConfig.Name == activeGW.Name
                if len(c.podCache) > 0 && (!isGateway || !c.podWorkersStarted) {
                        err := c.ofClient.UninstallMulticlusterPodFlows("")
			if err != nil {
				return err
			}
			c.podCache = map[string]*podNodeIPs{}
                }
``` 

The we handle Pod workers start/stop after the cleanup.

Reply:
Refined.